src/validators/validators.ts

```diff
@@ -19,6 +20,148 @@ export const withError = <T, Err1, Err2>(
   error: Err2
 ): Validator.Sync<T, Err2> => val => (validator(val) == null ? null : error);
 
+type Combine = {
+  <T, E, E1>(
+    validators: [Validator<T, E1>],
+    combinator: (errors: NullableTuple<[E1]>) => E
+  ): Validator<T, E>;
+  <T, E, E1, E2>(
+    validators: [Validator<T, E1>, Validator<T, E2>],
+    combinator: (errors: NullableTuple<[E1, E2]>) => E
+  ): Validator<T, E>;
+  <T, E, E1, E2, E3>(
+    validators: [Validator<T, E1>, Validator<T, E2>, Validator<T, E3>],
+    combinator: (errors: NullableTuple<[E1, E2, E3]>) => E
+  ): Validator<T, E>;
```

Comment:
couldn't get it working using `Variadic Tuple Types`

Reply:
```ts
type ValidatorsValue<Validators extends Validator<any, any>[]> =
  Validators extends Validator<infer Value, any>[] ? Value : never

type ValidatorsErrorsTuple<Validators extends Validator<any, any>[]> =
  Validators extends [...TypedValidationErrors<infer Err>] ? Err : never

type TypedValidationErrors<Errors extends readonly any[]> = {
  [Err in keyof Errors]: Validator<any, Errors[Err]>
}

declare const combineValidators: <E, Validators extends Validator<any, any>[]> (
  validators: [...Validators],
  combinator: (errors: ValidatorsErrorsTuple<Validators>) => E
) => Validator<ValidatorsValue<Validators>, E>;

// Example
declare const validator1: Validator<"one", "ERR_1">
declare const validator2: Validator<"one", "ERR_2">
declare const validator3: Validator<"one", "ERR_3">

const combined = combineValidators( // combined: Validator<"one", "ERR_COMBINED">
  [validator1, validator2, validator3],
  (err) => "ERR_COMBINED" as const // err: ["ERR_1", "ERR_2", "ERR_3"]
)
```
(hope i got the intention properly and didn't make any stupid mistake) Does this have a chance to work?

Reply:
nice, I didn't try this one :D

It kind of works, but it doesn't work as well as overload-hell version:
a) overload version does not compile at all when value type for one of the validators is not matching
b) type inference for inline (arrow function) validator value works with overload version, with this its `any`

So I think I'll stick with overlaods :(